File: tutorial/perl/BookmarkExample.pl
#!/usr/bin/env perl

use strict;
use warnings;

#Thrift components
use Thrift;
use Thrift::MemoryBuffer;
use Thrift::BinaryProtocol;
use Thrift::Socket;
use Thrift::FramedTransport;

#Bookmark
use lib "../gen-perl";
use Tutorial::Types;

#Thrudb
use Thrudex;
use Thrudoc;


package BookmarkManager;
use Time::HiRes qw(gettimeofday);
use Data::Dumper;

#Config
use constant THRUDEX_PORT   => 11299;
use constant THRUDOC_PORT    => 11291;

use constant THRUDOC_BUCKET  => "bookmarks";

use constant THRUDEX_DOMAIN => "tutorial";
use constant THRUDEX_INDEX  => "bookmarks";


sub new
{

    my $classname = shift;
    my $self      = {};

    #Used to serialize bookmark obj to string
    my $mbuf           = new Thrift::MemoryBuffer();
    my $mbuf_protocol  = new Thrift::BinaryProtocol($mbuf);

    $self->{mbuf}          = $mbuf;
    $self->{mbuf_protocol} = $mbuf_protocol;

    #Create object
    bless($self,$classname);

    #Connect to thrudb
    $self->connect_to_thrudoc();
    $self->connect_to_thrudex();

    return $self;
}

sub serialize_bookmark
{
    my $self = shift;
    my $b    = shift;

    my $mbuf          = $self->{mbuf};
    my $mbuf_protocol = $self->{mbuf_protocol};

    $mbuf->resetBuffer();

    $b->write($mbuf_protocol);

    return $mbuf->getBuffer();
}

sub deserialize_bookmark
{
    my $self  = shift;
    my $b_str = shift;
    my $b     = new Tutorial::Bookmark();

    my $mbuf          = $self->{mbuf};
    my $mbuf_protocol = $self->{mbuf_protocol};

    #set the buffer
    $mbuf->resetBuffer($b_str);

    $b->read($mbuf_protocol);

    return $b;
}

sub connect_to_thrudoc
{
    my $self = shift;

    eval{
       my $socket    = new Thrift::Socket("localhost",THRUDOC_PORT());
       my $transport = new Thrift::FramedTransport($socket);
       my $protocol  = new Thrift::BinaryProtocol($transport);
       $self->{thrudoc}  = new ThrudocClient($protocol);

       $transport->open();
    }; if($@) {
        die $@->{message} if UNIVERSAL::isa($@,"Thrift::TException");
        die $@;
    }
}

sub connect_to_thrudex
{
    my $self = shift;

    eval{
        my $socket    = new Thrift::Socket("localhost",THRUDEX_PORT());
        my $transport = new Thrift::FramedTransport($socket);
        my $protocol  = new Thrift::BinaryProtocol($transport);
        $self->{thrudex}  = new ThrudexClient($protocol);

        $transport->open();
    }; if($@){
        die $@->{message} if UNIVERSAL::isa($@,"Thrift::TException");
        die $@;
    }
}

sub load_tsv_file
{
    my $self = shift;
    my $file = shift;


    my $t0 = gettimeofday();

    open(BMFILE, "<$file") || die $!;

    while( chomp( my $line = <BMFILE> ) ){
        my($url,$title,$tags) = split("\t",$line);

        my $b = new Tutorial::Bookmark();

        $b->url  ( $url   );
        $b->title( $title );
        $b->tags ( $tags  );

        $self->add_bookmark($b);
    }

    close(BMFILE);

    $self->{thrudex}->commitAll();

    my $t1 = gettimeofday();

    print "*Indexed file in: ".($t1-$t0)."*\n\n";
}

sub add_bookmark
{
    my $self    = shift;
    my $b       = shift;

    die("Bad bookmark") unless defined $b;

    my $id = $self->store_bookmark($b);
    $self->index_bookmark($id,$b);
}

sub get_bookmark
{
    my $self    = shift;
    my $id      = shift;

    my $thrudoc = $self->{thrudoc};
    my $b_str   = $thrudoc->get(THRUDOC_BUCKET(),$id);

    return unless defined $b_str;

    return $self->deserialize_bookmark($b_str);
}

sub store_bookmark
{
    my $self     = shift;
    my $b        = shift;
    my $thrudoc  = $self->{thrudoc};

    my $b_str    = $self->serialize_bookmark($b);

    my $id = $thrudoc->putValue(THRUDOC_BUCKET(), $b_str );

    return $id;
}


sub index_bookmark
{
    my $self     = shift;
    my $thrudex = $self->{thrudex};

    my $id       = shift;
    my $b        = shift;

    #
    #Indexing requires mapping the fields in our
    #bookmark object to a Thrudex Document
    #
    my $doc      = new Thrudex::DocMsg();

    $doc->docid($id);
    $doc->domain( THRUDEX_DOMAIN() );
    $doc->index ( THRUDEX_INDEX()  );

    my $field = new Thrudex::Field();

    #
    #title
    #
    $field->name("title");
    $field->value($b->title);
    $field->sortable(1);
    push(@{$doc->{fields}}, $field);

    #
    #tags
    #
    $field = new Thrudex::Field();
    $field->name("tags");
    $field->value($b->tags);
    push(@{$doc->{fields}}, $field);

    $thrudex->add( $doc );
}

sub remove_all
{
    my $self     = shift;

    my $thrudoc  = $self->{thrudoc};
    my $thrudex  = $self->{thrudex};


    my $t0 = gettimeofday();

    #chunks of 100
    my $limit  = 100;
    my $r;
    my $seed;
    do{

        $r = $thrudoc->scan(THRUDOC_BUCKET(), $seed , $limit);

        if(@{$r->elements} > 0 ){

            my @thrudex_docs;

            foreach my $el (@{ $r->elements }){
                #
                my $rm = new Thrudex::RemoveMsg();
                $rm->domain( THRUDEX_DOMAIN() );
                $rm->index ( THRUDEX_INDEX()  );
                $rm->docid($el->key);

                push(@thrudex_docs, $rm);
            }

            $thrudex->removeList(\@thrudex_docs);
            $thrudoc->removeList($r->elements);

            $thrudex->commitAll();

        }

        $seed = $r->{seed};

    }while(@{$r->elements} == $limit);


    my $t1 = gettimeofday();

    print "\n*Index cleared in: ".($t1-$t0)."*\n";

}

sub find
{
    my $self  = shift;
    my $terms = shift;

    my $options = shift || {};

    return unless defined $terms;

    print "Searching for: '$terms' ".join(",",map{$_.":".$options->{$_}} keys %$options)."\n";
    my $t0 = gettimeofday();

    my $thrudex = $self->{thrudex};
    my $thrudoc  = $self->{thrudoc};
    my $query    = new Thrudex::QueryMsg();

    $query->domain( THRUDEX_DOMAIN() );
    $query->index(  THRUDEX_INDEX() );
    $query->query($terms);

    #$query->limit(10);
    #$query->offset(10);

    $query->randomize(1)               if exists $options->{random};
    $query->sortby($options->{sortby}) if exists $options->{sortby};

    my $ids = $thrudex->query( $query );

    return unless defined $ids;

    print "Found ".$ids->total." bookmarks\n";

    if(@{$ids->{ids}} > 0){

        my $response = $thrudoc->getList( $self->create_doc_list($ids->{ids}) );

        my @bookmarks;

        foreach my $r (@$response){

            if( $r->element->value ne "" ){

                my $bm = $self->deserialize_bookmark( $r->element->value );

                push(@bookmarks, $bm);
            }else{

                warn($r->ex->what);
            }

        }

        $self->print_bookmarks(\@bookmarks);
    }

    my $t1 = gettimeofday();
    print "Took: ".($t1-$t0)."\n\n";
}

sub create_doc_list
{
    my $self     = shift;
    my $ids      = shift;

    my @docs;

    foreach my $id (@$ids){
        my $el = new Thrudoc::Element();

        $el->bucket( THRUDOC_BUCKET() );
        $el->key( $id );

        push(@docs,$el);
    }

    return \@docs;
}

sub print_bookmarks
{
    my $self      = shift;
    my $bookmarks = shift;

    my $i = 1;
    foreach my $b (@$bookmarks){
        print "$i .\ttitle :\t".$b->title."\n";
        print "\turl   :\t(".$b->url.")\n";
        print "\ttags  :\t(".$b->tags.")\n";
        $i++;
    }
}

1;

#################################
#            Main
#################################
my $bm = new BookmarkManager();

my ($t0,$t1);
eval{

    $bm->load_tsv_file("../bookmarks.tsv");

    $bm->find("tags:(+css +examples)",{ random => 1});

    $bm->find("title:(linux)",{sortby => "title"});

    $bm->remove_all();

};if($@){
    die $@->{message} if UNIVERSAL::isa($@,"Thrift::TException");
    die $@;
}

File: tutorial/php/BookmarkExample.php
#!/usr/bin/env php
<?php

//Thrift Globals
$GLOBALS['THRIFT_ROOT'] = '/usr/lib/php5/thrift';

require_once $GLOBALS['THRIFT_ROOT'].'/Thrift.php';
require_once $GLOBALS['THRIFT_ROOT'].'/protocol/TBinaryProtocol.php';
require_once $GLOBALS['THRIFT_ROOT'].'/transport/TSocket.php';
require_once $GLOBALS['THRIFT_ROOT'].'/transport/TFramedTransport.php';
require_once $GLOBALS['THRIFT_ROOT'].'/transport/TMemoryBuffer.php';

error_reporting(E_NONE);
$GEN_DIR = '../gen-php';
require_once $GEN_DIR.'/Thrudex.php';
require_once $GEN_DIR.'/Thrudex_types.php';
require_once $GEN_DIR.'/Thrudoc.php';
require_once $GEN_DIR.'/Thrudoc_types.php';
require_once $GEN_DIR.'/Bookmark_types.php';
error_reporting(E_ALL);


define("THRUDEX_PORT", 11299 );
define("THRUDOC_PORT",  11291 );

define("THRUDOC_BUCKET", "bookmarks");

define("THRUDEX_INDEX",  "bookmarks");

class BookmarkManager {
    private $thrudoc;
    private $thrudex;

    private $mbuf;
    private $mbuf_protocol;

    function __construct() {
        $this->connect_to_thrudoc();
        $this->connect_to_thrudex();

        $this->mbuf          = new TMemoryBuffer();
        $this->mbuf_protocol = new TBinaryProtocol($this->mbuf);
    }

    private function connect_to_thrudoc()
    {
        //Thrudoc connection
        $socket         = new TSocket('localhost', THRUDOC_PORT);
        $transport      = new TFramedTransport($socket);
        $protocol       = new TBinaryProtocol($transport);
        $this->thrudoc  = new ThrudocClient($protocol);

        $transport->open();
    }

    private function connect_to_thrudex()
    {
        //Thrudex connection
        $socket         = new TSocket('localhost', THRUDEX_PORT);
        $transport      = new TFramedTransport($socket);
        $protocol       = new TBinaryProtocol($transport);
        $this->thrudex = new ThrudexClient($protocol);

        $transport->open();

        $this->thrudex->admin("create_index", THRUDEX_INDEX);
    }

    private function serialize_bookmark($b)
    {

        //Convert object to string
        $b->write($this->mbuf_protocol);

        $b_str = $this->mbuf->read( $this->mbuf->available() );


        return $b_str;
    }

    private function deserialize_bookmark($b_str)
    {
        $b = new Tutorial_Bookmark();

        //set the buffer
        $this->mbuf->write($b_str);

        $b->read($this->mbuf_protocol);

        return $b;
    }

    function load_tsv_file($file)
    {
        $t0 = gettimeofday(true);

        $fh = fopen($file, 'r');

        while( $line = fgets($fh) ){
            $arr = split("\t", $line);

            $b = new Tutorial_Bookmark();
            $b->url   = $arr[0];
            $b->title = $arr[1];
            $b->tags  = $arr[2];

            $this->add_bookmark($b);
        }

        fclose($fh);

        $t1 = gettimeofday(true);

        print "*Indexed file in: ".($t1-$t0)."*\n\n";
    }

    private function add_bookmark( $b )
    {
        $id = $this->store_bookmark($b);
        $this->index_bookmark($id,$b);
    }

    private function get_bookmark($id)
    {
        $b_str = $this->thrudoc->fetch($id);
        return $this->deserialize_bookmark($b_str);
    }

    private function store_bookmark( $b )
    {
        $b_str = $this->serialize_bookmark($b);
        $id    = $this->thrudoc->putValue(THRUDOC_BUCKET,$b_str);

        return $id;
    }

    private function index_bookmark( $id, $b )
    {
        $doc      = new Thrudex_Document();

        $doc->key    = $id;
        $doc->index  = THRUDEX_INDEX;

        $fields = array();
        $field = new Thrudex_Field();

        //
        //title
        //
        $field->key      = "title";
        $field->value    = $b->title;
        $field->sortable = true;
        array_push( $fields, $field);


        $field = new Thrudex_Field();
        //
        //tag
        //
        $field->key   = "tags";
        $field->value = $b->tags;
        array_push( $fields, $field);

        $doc->fields = $fields;

        $this->thrudex->put( $doc );
    }

    function remove_all()
    {
        $t0 = gettimeofday(true);

        //chunks of 100
        $limit  = 100;
        $seed   = "";
        do{

            $r    = $this->thrudoc->scan(THRUDOC_BUCKET,$seed,$limit);
            $docs = array();

            if(count($r->elements) == 0)
                break;

            foreach($r->elements as $id){
                $el = new Thrudex_Element();
                $el->index  = THRUDEX_INDEX;
                $el->key    = $id->key;

                array_push( $docs, $el );
            }

             $this->thrudex->removeList($docs);
             $this->thrudoc->removeList($r->elements);

             $seed = $r->seed;

        }while( count($r->elements) >= $limit );


        $t1 = gettimeofday(true);
        print "\n*Index cleared in: ".($t1-$t0)."*\n";
    }

    function find($terms, $options)
    {
        $func = create_function('$x,$y', 'return "$x:$y";');

        print "Searching for: '$terms' ".join(",",array_map($func, array_keys($options), array_values($options)))."\n";

        $t0 = gettimeofday(true);

        $query    = new Thrudex_SearchQuery();

        $query->index  =  THRUDEX_INDEX;
        $query->query  = $terms;

        //$query->limit = 10;
        //$query->offset= 10;

        if( array_key_exists( "random", $options ) )
            $query->randomize = true;

        if( array_key_exists( "sortby", $options ) )
            $query->sortby    = $options["sortby"];


        $els = $this->thrudex->search( $query );

        print "Found ".$els->total." bookmarks\n";

        if( count( $els->elements ) > 0 ){
            $response  = $this->thrudoc->getList( $this->create_doc_list($els->elements) );
            $bookmarks = array();

            foreach( $response as $doc ){
                if( $doc->element->value != null )
                    array_push( $bookmarks, $this->deserialize_bookmark($doc->element->value) );
                else
                    print "Could not get Document: \n";
            }

            $this->print_bookmarks($bookmarks);
        }

        $t1 = gettimeofday(true);

        print "Took: ".($t1-$t0)."\n\n";
    }

    private function create_doc_list( $ids )
    {
        $docs = array();

        foreach( $ids as $id ){
            $doc = new Thrudoc_Element();

            $doc->key    = $id->key;
            $doc->bucket = THRUDOC_BUCKET;

            array_push( $docs, $doc );
        }

        return $docs;
    }

    private function print_bookmarks( $bookmarks )
    {
        $i = 1;

        foreach($bookmarks as $b){
            print "$i .\ttitle :\t".$b->title."\n";
            print "\turl   :\t(".$b->url.")\n";
            print "\ttags  :\t(".$b->tags.")\n";
            $i++;
        }

    }
}

//################################
//#            Main
//################################

$bm = new BookmarkManager();

try{

    $bm->load_tsv_file("../bookmarks.tsv");

    $bm->find("tags:(+css +examples)",array( "random" => true));

    $bm->find("title:(linux)",array("sortby" => "title"));

    $bm->remove_all();

}catch(TException $tx) {
  print 'TException: '.$tx->getMessage()."\n";
}

?>
